pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -1562,6 +1562,14 @@ class Timestamp(_Timestamp):
             _TSObject ts
             tzinfo_type tzobj
 
+        # GH#52117 If we passed positional args, then _ts_input
+        # is now set to year and the other positional args
+        # are shifted to the left. Let's shift back
+        if (isinstance(ts_input, int) and
+                isinstance(year, int) and isinstance(month, int)):
+            ts_input, year, month, day, hour, minute, second, microsecond = (
```

Comment:
Would this be incorrect if a user passed `Timestamp(2023, 1, 1, minute=1)` by assigning minute to second?

Reply:
Thanks, this is a good point. We get `Timestamp('2023-01-01 00:00:01')`. 

I hate it, but should I roll back to the bugfix version where I didn't touch the args, but instead just checked whether the first three args were filled or not and prevented the raise?

```python
# GH#52117 note that currently we end up with the year in the ts_input
            # variable when using positional arguments. We should fix this.
            from_positional = (ts_input is not None and
                               year is not None and month is not None)

            if (ts_input is not _no_input and not from_positional and not (
                    PyDateTime_Check(ts_input) and
                    getattr(ts_input, "tzinfo", None) is None)):
                raise ValueError(
                    "Cannot pass fold with possibly unambiguous input: int, "
                    "float, numpy.datetime64, str, or timezone-aware "
                    "datetime-like. Pass naive datetime-like or build "
                    "Timestamp from components."
                )
```

Another solution would be to make all the args up to second positional-only, since Python 3.8 allows that, but it would deviate from the datetime API and be a breaking change. That would need a separate PR and a deprecation cycle, I think.

Reply:
Hm, there might be a different solution, give me a minute.

Reply:
Yeah I would suggest tackling the init argument complexity in a follow up unless it's a truly a blocker for this fix

Reply:
@mroeschke Thanks for giving that example. Turns out, there is another bug caused by the same anomaly in the data. We currently have on main:

```python
pd.Timestamp(2020, 12, 31, second=1)

> Timestamp('2020-12-31 00:00:00.000001')
```

 This looks like a bug to me. In addition, we are testing for this unexpected behavior [here](https://github.com/pandas-dev/pandas/blob/d86cdb0c230b311036d154312448de87e59a6c8b/pandas/tests/scalar/timestamp/test_constructors.py#L341):
```python
    @pytest.mark.parametrize("kwd", ["nanosecond", "microsecond", "second", "minute"])
    def test_constructor_positional_keyword_mixed_with_tzinfo(self, kwd, request):
        # TODO: if we passed microsecond with a keyword we would mess up
        #  xref GH#45307
        if kwd != "nanosecond":
            # nanosecond is keyword-only as of 2.0, others are not
            mark = pytest.mark.xfail(reason="GH#45307")
            request.node.add_marker(mark)

        kwargs = {kwd: 4}
        ts = Timestamp(2020, 12, 31, tzinfo=timezone.utc, **kwargs)

        td_kwargs = {kwd + "s": 4}
        td = Timedelta(**td_kwargs)
        expected = Timestamp("2020-12-31", tz=timezone.utc) + td
        assert ts == expected
```

My proposed fix makes it so that the fails here no longer happen and that:
```python
pd.Timestamp(2020, 12, 31, second=1)

> Timestamp('2020-12-31 00:00:01')
```

Reply:
> Yeah I would suggest tackling the init argument complexity in a follow up unless it's a truly a blocker for this fix

That's one option but please reveiew the current proposed solution. I think it's not terribly costly and makes the constructor easier to maintain and less vulnerable to bugs.